src/gt4py/next/program_processors/runners/dace_iterator/itir_to_tasklet.py

```diff
@@ -625,47 +724,156 @@ def _visit_indirect_addressing(self, node: itir.FunCall) -> IteratorExpr:
+            init_tasklet = init_state.add_tasklet(
+                "init_reduce", {}, {"__out"}, f"__out = {init_value}"
+            )
+            init_state.add_edge(
+                init_tasklet,
+                "__out",
+                init_state.add_access(result_name),
+                None,
+                dace.Memlet.simple(result_name, "0"),
+            )
+
+            # set reduction state to enable dereference of neighbors in input fields and to set WCR on reduce tasklet
+            self.context.reduce_limit = nreduce
+            self.context.reduce_wcr = "lambda x, y: " + _MATH_BUILTINS_MAPPING[str(op_name)].format(
+                "x", "y"
+            )
+
+            # visit child nodes for input arguments
+            for i, node_arg in enumerate(node.args):
+                if not args[i]:
+                    args[i] = self.visit(node_arg)[0]
+
+            lambda_node = itir.Lambda(expr=fun_node.expr.args[1], params=fun_node.params[1:])
+            lambda_context, inner_inputs, inner_outputs = self.visit(lambda_node, args=args)
+
+            # clear context
+            self.context.reduce_limit = 0
+            self.context.reduce_wcr = None
+
+            # the connectivity arrays (neighbor tables) are not needed inside the reduce lambda SDFG
+            neighbor_tables = filter_neighbor_tables(self.offset_provider)
+            for conn, _ in neighbor_tables:
+                var = connectivity_identifier(conn)
+                lambda_context.body.remove_data(var)
+            # cleanup symbols previously used for shape and stride of connectivity arrays
+            p = RemoveUnusedSymbols()
+            p.apply_pass(lambda_context.body, {})
+
+            input_memlets = [
+                create_memlet_at(expr.value.data, ("__idx",)) for arg, expr in zip(node.args, args)
+            ]
+            output_memlet = dace.Memlet.simple(result_name, "0")
+
+            input_mapping = {param: arg for (param, _), arg in zip(inner_inputs, input_memlets)}
+            output_mapping = {inner_outputs[0].value.data: output_memlet}
+            symbol_mapping = map_nested_sdfg_symbols(
+                self.context.body, lambda_context.body, input_mapping
+            )
+
+            nsdfg_node, map_entry, _ = add_mapped_nested_sdfg(
+                self.context.state,
+                sdfg=lambda_context.body,
+                map_ranges=nreduce_domain,
+                inputs=input_mapping,
+                outputs=output_mapping,
+                symbol_mapping=symbol_mapping,
+                input_nodes={arg.value.data: arg.value for arg in args},
+                output_nodes={result_name: result_access},
+            )
+
+            # map fusion is beneficial for reduction stencils:
+            # the solution was designed to keep code generation simple and let DaCe fuse maps
+            self.context.body.apply_transformations_repeated([MapFusion], validate=False)
```

Comment:
I don't think it's a good idea to put any optimization/canonicalization transforms in here unless you deliberately generate IR that doesn't meet post-conditions and you fix that with already existing passes.

Reply:
I think this is the case, as I wrote in the comment. Consider this reduction SDFG without applying `MapFusion`:
![image](https://github.com/GridTools/gt4py/assets/25990065/c3124dc6-3170-48c2-ab0f-6c505504e6cc)

After applying `MapFusion` we get a simplified SDFG:
![image](https://github.com/GridTools/gt4py/assets/25990065/038f2be4-ffc6-4241-9435-721c5a5d50d6)

Reply:
Hmm, so what I meant is for example if you already have a compiler pass that converts AVX instruction to SSE, it may be reasonable to write only the AVX vectorization pass, and that let's you target both ISAs together with the conversion pass. In this case, when you vectorize with SSE, emitting AVX instructions violates the post conditions, so you shouldn't do that, so you first vectorize with AVX, and then call the AVX to SSE conversion.

In your case, you first render the reductions, and then call map fusion. My question is, if you don't call map fusion, will your output have invalid "instructions" (like AVX)? If not, like I think is the case for you, it's probably not a wise idea to run optimization passes. Instead, you can just leave the generated (but still valid) IR as unoptimized as you want, and then once you're completely done with code generation you can run optimization passes over the entire IR, in a well-thought out order, with access to global information.

Reply:
I get your point now, to prefer optimization with global knowledge, rather than doing it early inside the nested-SDFG. It is a good point. My intention was to run this transformation here as a cleanup pass rather than an optimization pass. In addition to the example above, there is another use case where we always generate redundant maps, that is deferencing with the neighbor index (`test_external_local_field`). So it is 2 known cases of redundant SDFG code.
In principle,  I would like not to generate redundant maps but this is the result of ITIR representation and traversal of reduction.

Reply:
I can remove the `MapFusion` transformation. As you say, the SDFG is valid in any case and produces functionally correct code. We'll postpone it to a future optimization phase with global knowledge.

Reply:
You're only running it localized to the part of the SDFG that's generated for the reduction, and not the whole SDFG, right? I think in that case it's okay to run the fusion, but can you perhaps comment that it's for cosmetic purposes, so that both the SDFG generating code is simple and the generated SDFG is simple.